File: orchestrator/review.py
from dataclasses import dataclass
from typing import Any

from orchestrator import context, session, log

logger = log.get("review")


@dataclass
class ReviewResult:
    returncode: int
    stdout: str
    stderr: str


def _build_review_prompt(
    review_type: str,
    project_name: str | None = None,
    milestone: dict[str, Any] | None = None,
    handoff_notes: list[Any] | None = None,
    verify_results: list[Any] | None = None,
) -> tuple[str, str]:
    milestone_list = [milestone] if milestone else []
    bundle = context.build_bundle(
        milestones=milestone_list,
        handoff_notes=handoff_notes or [],
        verify_results=verify_results or [],
    )

    if context.exceeds_threshold(bundle):
        logger.warning("Context bundle exceeds token budget, degrading")
        degraded = False
        if verify_results:
            bundle = context.degrade(bundle, strategy="no_verify")
            degraded = True
        if handoff_notes and context.exceeds_threshold(bundle):
            bundle = context.degrade(bundle, strategy="no_handoff")
            degraded = True
        if not degraded:
            bundle = context.degrade(bundle, strategy="minimal")

    target = project_name or (milestone or {}).get("name", "unknown")
    prompt = (
        f"Review: {review_type} for {target}\n\n"
        f"## Context Bundle\n\n{bundle}\n\n"
        f"Please perform a {review_type} review and provide feedback. "
        f"Write your review to HANDOFF.md in the current directory."
    )
    return bundle, prompt


def review_phase(
    milestone: dict[str, Any],
    handoff_notes: list[Any] | None = None,
    verify_results: list[Any] | None = None,
) -> ReviewResult:
    bundle, prompt = _build_review_prompt(
        review_type="phase",
        milestone=milestone,
        handoff_notes=handoff_notes,
        verify_results=verify_results,
    )
    logger.info("starting phase review for milestone %r", milestone.get("name"))
    result = session.run(prompt)
    return ReviewResult(
        returncode=result.returncode,
        stdout=result.stdout,
        stderr=result.stderr,
    )


def review_final(
    project_name: str,
    milestones: list[dict[str, Any]],
    handoff_notes: list[Any] | None = None,
    verify_results: list[Any] | None = None,
) -> ReviewResult:
    bundle, prompt = _build_review_prompt(
        review_type="final",
        project_name=project_name,
        milestone=None,
        handoff_notes=handoff_notes,
        verify_results=verify_results,
    )
    logger.info("starting final review for project %r", project_name)
    result = session.run(prompt)
    return ReviewResult(
        returncode=result.returncode,
        stdout=result.stdout,
        stderr=result.stderr,
    )

File: orchestrator/internal/review/review.go
package review

import (
	"fmt"
	"time"

	"orchestrator/internal/context"
	"orchestrator/internal/handoff"
	"orchestrator/internal/session"
	"orchestrator/internal/verify"
)

type Result struct {
	ReturnCode int
	Stdout     string
	Stderr     string
}

func buildReviewPrompt(reviewType string, projectName string, milestones []context.MilestoneInfo, handoffNotes []handoff.Note, verifyResults []verify.Result) (string, string) {
	bundle := context.BuildBundle(milestones, handoffNotes, verifyResults)

	if context.ExceedsThreshold(bundle, 0) {
		degraded := false
		if len(verifyResults) > 0 {
			bundle = context.Degrade(bundle, context.DegradeNoVerify)
			degraded = true
		}
		if len(handoffNotes) > 0 && context.ExceedsThreshold(bundle, 0) {
			bundle = context.Degrade(bundle, context.DegradeNoHandoff)
			degraded = true
		}
		if !degraded {
			bundle = context.Degrade(bundle, context.DegradeMinimal)
		}
	}

	target := projectName
	prompt := fmt.Sprintf(
		"Review: %s for %s\n\n## Context Bundle\n\n%s\n\nPlease perform a %s review and provide feedback. Write your review to HANDOFF.md in the current directory.",
		reviewType, target, bundle, reviewType,
	)
	return bundle, prompt
}

func Phase(milestoneName, milestoneSpec string, handoffNotes []handoff.Note, verifyResults []verify.Result) (*Result, error) {
	ms := []context.MilestoneInfo{{Name: milestoneName, Spec: milestoneSpec}}
	_, prompt := buildReviewPrompt("phase", milestoneName, ms, handoffNotes, verifyResults)
	result, err := session.Run(prompt, 5*time.Minute)
	if err != nil {
		return nil, err
	}
	return &Result{
		ReturnCode: result.ReturnCode,
		Stdout:     result.Stdout,
		Stderr:     result.Stderr,
	}, nil
}

func Final(projectName string, milestones []context.MilestoneInfo, handoffNotes []handoff.Note, verifyResults []verify.Result) (*Result, error) {
	_, prompt := buildReviewPrompt("final", projectName, milestones, handoffNotes, verifyResults)
	result, err := session.Run(prompt, 5*time.Minute)
	if err != nil {
		return nil, err
	}
	return &Result{
		ReturnCode: result.ReturnCode,
		Stdout:     result.Stdout,
		Stderr:     result.Stderr,
	}, nil
}
